epuck_ros2_cpp/src/driver.cpp

```diff
@@ -0,0 +1,447 @@
+      dist[2],       // -90
+      OUT_OF_RANGE,  // -75
+      OUT_OF_RANGE,  // -60
+      dist[1],       // -45
+      OUT_OF_RANGE,  // -30
+      dist[0],       // -15
+      OUT_OF_RANGE,  // dist['tof'], // 0
+      dist[7],       // 15
+      OUT_OF_RANGE,  // 30
+      dist[6],       // 45
+      OUT_OF_RANGE,  // 60
+      OUT_OF_RANGE,  // 75
+      dist[5],       // 90
+      OUT_OF_RANGE,  // 105
+      OUT_OF_RANGE,  // 120
+      OUT_OF_RANGE,  // 135
+      dist[4],       // 150
+    };
+    laser_publisher->publish(msg);
+
+    // Create Range messages
+    for (int i = 0; i < 8; i++) {
+      auto msg_range = sensor_msgs::msg::Range();
+      msg_range.header.stamp = stamp;
+      msg_range.header.frame_id = "ps" + std::to_string(i);
+      msg_range.radiation_type = sensor_msgs::msg::Range::INFRARED;
+      msg_range.min_range = 0.05 + SENSOR_DIST_FROM_CENTER;
+      msg_range.min_range = 0.005 + SENSOR_DIST_FROM_CENTER;
+      msg_range.range = dist[i];
+      range_publisher[i]->publish(msg_range);
+    }
+  }
+
+  void publish_odometry_data(rclcpp::Time &stamp) {
+    const int16_t left_wheel_raw = (msg_sensors[41] & 0x00FF) | ((msg_sensors[42] << 8) & 0xFF00);
+    const int16_t right_wheel_raw = (msg_sensors[43] & 0x00FF) | ((msg_sensors[44] << 8) & 0xFF00);
+
+    // Handle overflow
+    // The MCU can handle only 2 bytes of ticks (about 4m), therefore this code allows us to
+    // track even when the number of ticks has reached maximum value of 2^15. It is based on
+    // detecting the overflow and updating counter `odom_left_overflow`/`odom_right_overflow`.
+    const int64_t prev_left_wheel_corrected = odom_left_overflow * POW2(16) + prev_left_wheel_raw;
+    const int64_t prev_right_wheel_corrected = odom_right_overflow * POW2(16) + prev_right_wheel_raw;
+    if (abs((int64_t)prev_left_wheel_raw - (int64_t)left_wheel_raw) > POW2(15) - ODOM_OVERFLOW_GRACE_TICKS)
+      odom_left_overflow = (prev_left_wheel_raw > 0 && left_wheel_raw < 0) ? odom_left_overflow + 1 : odom_left_overflow - 1;
+    if (abs((int64_t)prev_right_wheel_raw - (int64_t)right_wheel_raw) > POW2(15) - ODOM_OVERFLOW_GRACE_TICKS)
+      odom_right_overflow =
+        (prev_right_wheel_raw > 0 && right_wheel_raw < 0) ? odom_right_overflow + 1 : odom_right_overflow - 1;
+    const int64_t left_wheel_corrected = odom_left_overflow * POW2(16) + left_wheel_raw;
+    const int64_t right_wheel_corrected = odom_right_overflow * POW2(16) + right_wheel_raw;
+    const float left_wheel_rad = left_wheel_corrected / (ENCODER_RESOLUTION / (2 * M_PI));
+    const float right_wheel_rad = right_wheel_corrected / (ENCODER_RESOLUTION / (2 * M_PI));
+    const float prev_left_wheel_rad = prev_left_wheel_corrected / (ENCODER_RESOLUTION / (2 * M_PI));
+    const float prev_right_wheel_rad = prev_right_wheel_corrected / (ENCODER_RESOLUTION / (2 * M_PI));
+
+    // Calculate velocities
+    const float v_left_rad = (left_wheel_rad - prev_left_wheel_rad) / PERIOD_S;
```

Comment:
I am wondering if using this fixed period is safer enough.
Wouldn't it be safer to measure exact time instead?

Reply:
Hmm, that is a good point. On the one hand:
- `.create_wall_timer()` should provide very precise intervals (of course, it depends on the rest of the code in the callback, but the execution time is nearly the same) and
- as far as the sampling period is small we should get quite good results.
but:
- this line https://github.com/cyberbotics/epuck_ros2/blob/ae5f05ac32bc0edc60bef58cc2cd7005e47aa1cc/epuck_ros2_cpp/src/driver.cpp#L409 can fail and the sampling period is comprised, (although this almost never should occur and the error should not be big).

If we aim for highly precise odometry we should do the following:
- replace `PERIOD_S` with the measured time difference,
- replace float variables to double and
- increase sampling frequency.
Of course, this will cause performance issues.

Reply:
My main concern was that the call of `publish_odometry_data(stamp); ` is encapsulated in a if letting think that it might not be called at each step, but if this is extremely rare than it's probably ok.

Reply:
It checks for I2C failures now and corrects sample rate accordingly